Enable implicit operator versioning via number of arguments

iseeyuan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58852 Enable implicit operator versioning via number of arguments**

Enable implicit operator versioning via number of arguments from Mobile.
1. By default, TS doesn't emit instructions for tailing default args and the provided number of specified args is serialized to bytecode. From interpreter the default values are fetched from operator schema. The implementation has been landed in #56845. Please refer to #56845 for details.
2. Since there is bytecode schema change, the bytecode version is bumped from 5 to 6.
3. The corresponding backport function is provided, for forward compatibility use. Note that because there is instruction change, a global flag is used as the switch to control the two versions.

Differential Revision: [D28789746](https://our.internmc.facebook.com/intern/diff/D28789746)

iseeyuan committed:
Enable implicit operator versioning via number of arguments

[ghstack-poisoned]

iseeyuan committed:
Update on "Enable implicit operator versioning via number of arguments"




[ghstack-poisoned]

cccclai commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -847,5 +853,16 @@ std::vector<std::string> export_opnames(const script::Module& m) {
 
+// Thread local floag (only happens in export, i.e. on server side)
+// to control if
+thread_local bool emitBytecodeDefaultInputs =
```
Looks like it can be fixed.

iseeyuan replied:
This value will be changed in the backport function, so directly apply const does not work. Do you have any suggestion?

cccclai replied:
Hmm I see. Probably try suppress this message, otherwise clang-tidy will always complains this line in CI

cccclai commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -627,12 +625,20 @@ void ScriptModuleSerializer::writeByteCode(
+  // backport function, change version number in both bytecode.pkl,
+  // and other files containling the version number, like
+  // mobile_debug_handles.pkl.
+  if (version_to_write == 6 && BytecodeEmitDefaultInputsMode::is_enabled()) {
```
Ideally it would be good to read the bytecode.pkl (v6) and just rewrite the version number to v5. If it's urgent, let's follow up on it.

raziel replied:
yeah this should be part of the backport code not in the serialization code.
Basically the output of a backport function should be a model in the previous version, and all the logic to make that happen that's versioning specific (like setting the right version)  should be self contained in the function.

gmagogsfm replied:
+1 on what Raziel said.

Also, it seems dangerous to check `version_to_write` against magic number `6`. In the future, when byte code version bumps to 7, this code will not be triggered correctly.

iseeyuan replied:
Thanks! I'll make a util function just to alter the version number.

iseeyuan replied:
I have tried a few options (https://fb.quip.com/MNuMAOgiHQdT), but have not succeeded in just updating the version number in bytecode.pkl. The fundamental reason is that we rely on writeArchive() to write the .pkl file, which is stateful. It depends on the constant metadata which was written before, which is related to the tensor's storageImple pointer. It's different from time to time for each write. 

It's also coupled with another state, type_name_uniquer_ to resolve type name collisions.

dhruvbird replied:
@iseeyuan Updated the document at https://fb.quip.com/MNuMAOgiHQdT#DVLACAoItek with some py code to update just the bytecode in the pickle. Can be done in C++ too.

cccclai commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -212,6 +213,35 @@ bool backport_v5_to_v4(
+  // resolved at runtime init stage for better operator compatibility.
+  BytecodeEmitDefaultInputsGuard argNumGuard(true);
+  std::stringstream output;
+  torch_script._save_for_mobile(output, extra_files, hasBytecodeDebug);
```
Hmm seems like `extra_files` is declared in line227. How does it preserved?

raziel replied:
It gets populated here:
```
Module torch_script = torch::jit::load(rai, c10::nullopt, extra_files);
```
Then saved back in 

```
torch_script._save_for_mobile(output, extra_files, hasBytecodeDebug);
```

dhruvbird replied:
@cccclai  has a point (I think). If you see the code for deserialize https://www.internalfb.com/intern/diffusion/FBS/browse/master/xplat/caffe2/torch/csrc/jit/serialization/import.cpp?commit=87aa02c6427125d488d5688a9bee69ee1b75f163&lines=237-253

it suggests that one needs to pass in the list of extra files to read. In this case the passed in map `extra_files` is empty, so no extra files will be read in, and hence no extra files will be preserved.

dhruvbird replied:
@iseeyuan maybe I'm missing some details, and hence the question. Q. Will this line write out the model based on the current bytecode version and not v5?

iseeyuan replied:
@dhruvbird It was intended to write v5 format bytecode.

cccclai reviewed (commented):
LGTM in general.

raziel commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -212,6 +213,35 @@ bool backport_v5_to_v4(
+  BytecodeEmitDefaultInputsGuard argNumGuard(true);
+  std::stringstream output;
+  torch_script._save_for_mobile(output, extra_files, hasBytecodeDebug);
+  std::stringstream output_copy(output.str());
```
Shouldn't this code

```
output_model_stream.swap(output);
```

use output_copy instead?

raziel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -847,5 +853,16 @@ std::vector<std::string> export_opnames(const script::Module& m) {
 }
 
+// Thread local floag (only happens in export, i.e. on server side)
+// to control if
```
to control if ...?

maybe add more detail what it controls.

iseeyuan replied:
Will update it. Thanks!

raziel reviewed (changes requested):
Thanks Martin, there's a few comments to address.

gmagogsfm commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -212,6 +213,35 @@ bool backport_v5_to_v4(
+  // the instructions (LOADC, for example), to push the values to the stack. It
+  // restores the behavior of V5 and before. For V6, the default arg values are
+  // resolved at runtime init stage for better operator compatibility.
+  BytecodeEmitDefaultInputsGuard argNumGuard(true);
```
Nit:

Place `argNumGuard` in a tighter scope to avoid accidental leak of guard effects

gmagogsfm commented on torch/csrc/jit/serialization/export.h:
```diff
@@ -188,5 +188,21 @@ TORCH_API void SetExportModuleMobileInfoConverter(
  */
 TORCH_API std::vector<std::string> export_opnames(const Module& m);
 
+struct TORCH_API BytecodeEmitDefaultInputsMode {
```
```suggestion
struct TORCH_API BytecodeEmitDefaultValueForUnspecifiedArgMode {
```

Please also consider adding comment explaining what this guard is doing for future readers

dhruvbird replied:
+1 What does having this mode armed or disarmed mean?

iseeyuan replied:
@dhruvbird It's used to control if the TorchScript function emits the default value for unspecified argument or not.

gmagogsfm commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -212,6 +213,35 @@ bool backport_v5_to_v4(
   return true;
 }
 
+bool backport_v6_to_v5(
```
OOC, is there a finite window for keeping this backport functionality?

iseeyuan replied:
It's from current to 4.

dhruvbird commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -212,6 +213,35 @@ bool backport_v5_to_v4(
   return true;
 }
 
+bool backport_v6_to_v5(
```
Q. Should we consider using an exception to indicate a failure to backport? Mentioning this since it is possible for someone to forget to check the return value of this method since the compiler can't check the `[[noreturn]]` annotation on functions called via pointer.

cccclai replied:
In #58790, I add [a few check in the main while loop to catch backport failure](https://github.com/pytorch/pytorch/blob/03d1b0653fb905313c22e3e1e2a2a67140f309ef/torch/csrc/jit/mobile/backport_manager.cpp#L287-L328). Since this api is used in different model delivery infra and it's better to gracefully exit, I tend to use warning message + boolean to indicate the backport failure.

dhruvbird replied:
imho, as long as the caller is able to catch the exception, it should be okay. In fact it maybe safer to use exceptions, since the risk of accidentally forgetting to check the return value goes away, and the call-site makes it super clear that an exception is expected. We could consider this as a future improvement to make the codebase safer from accidental errors resulting from not checking the return value (which is the intention behind the `[[noreturn]]` annotation in C++).

dhruvbird replied:
I'm seeing this code in the link you referenced above:

<!--StartFragment-->
    // Keep backporting till request version
    std::stringstream backport_model_stream =
        bytecodeBackportFunctions()[bytecode_version--](input_model_stream);
<!--EndFragment-->

Has the API changed since then?

cccclai commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -212,6 +213,35 @@ bool backport_v5_to_v4(
+bool backport_v6_to_v5(
+    std::stringstream& input_model_stream,
+    std::stringstream& output_model_stream) {
+  std::shared_ptr<IStreamAdapter> rai =
+      std::make_shared<IStreamAdapter>(&input_model_stream);
+  auto reader = std::make_shared<PyTorchStreamReader>(rai);
```
Actually, since reader is not used anywhere else, we probably don't a `share_ptr`:
```
auto reader = PyTorchStreamReader(&input_model_stream);
```
Accordingly, change `reader->hasRecord` to `reader.hasRecord`.

tugsbayasgalan commented on torch/csrc/jit/serialization/export.h:
```diff
@@ -188,5 +188,21 @@ TORCH_API void SetExportModuleMobileInfoConverter(
+  static void set_enabled(bool enabled);
+};
+
+struct TORCH_API BytecodeEmitDefaultInputsGuard {
```
why do we need to have a guard around this flag?

iseeyuan committed:
Update on "Enable implicit operator versioning via number of arguments"


Enable implicit operator versioning via number of arguments from Mobile.
1. By default, TS doesn't emit instructions for tailing default args and the provided number of specified args is serialized to bytecode. From interpreter the default values are fetched from operator schema. The implementation has been landed in #56845. Please refer to #56845 for details.
2. Since there is bytecode schema change, the bytecode version is bumped from 5 to 6.
3. The corresponding backport function is provided, for forward compatibility use. Note that because there is instruction change, a global flag is used as the switch to control the two versions. 


[ghstack-poisoned]

iseeyuan committed:
Update on "Enable implicit operator versioning via number of arguments"


Enable implicit operator versioning via number of arguments from Mobile.
1. By default, TS doesn't emit instructions for tailing default args and the provided number of specified args is serialized to bytecode. From interpreter the default values are fetched from operator schema. The implementation has been landed in #56845. Please refer to #56845 for details.
2. Since there is bytecode schema change, the bytecode version is bumped from 5 to 6.
3. The corresponding backport function is provided, for forward compatibility use. Note that because there is instruction change, a global flag is used as the switch to control the two versions. 


[ghstack-poisoned]

iseeyuan committed:
Update on "Enable implicit operator versioning via number of arguments"


Enable implicit operator versioning via number of arguments from Mobile.
1. By default, TS doesn't emit instructions for tailing default args and the provided number of specified args is serialized to bytecode. From interpreter the default values are fetched from operator schema. The implementation has been landed in #56845. Please refer to #56845 for details.
2. Since there is bytecode schema change, the bytecode version is bumped from 5 to 6.
3. The corresponding backport function is provided, for forward compatibility use. Note that because there is instruction change, a global flag is used as the switch to control the two versions. 


[ghstack-poisoned]

iseeyuan commented:
@iseeyuan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28789746).

iseeyuan committed:
Update on "Enable implicit operator versioning via number of arguments"


Enable implicit operator versioning via number of arguments from Mobile.
1. By default, TS doesn't emit instructions for tailing default args and the provided number of specified args is serialized to bytecode. From interpreter the default values are fetched from operator schema. The implementation has been landed in #56845. Please refer to #56845 for details.
2. Since there is bytecode schema change, the bytecode version is bumped from 5 to 6.
3. The corresponding backport function is provided, for forward compatibility use. Note that because there is instruction change, a global flag is used as the switch to control the two versions.

Differential Revision: [D28789746](https://our.internmc.facebook.com/intern/diff/D28789746)

[ghstack-poisoned]

iseeyuan commented:
@iseeyuan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28789746).

cccclai commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -211,6 +212,152 @@ std::stringstream backport_v5_to_v4(std::stringstream& input_model_stream) {
+  std::string fname = archive_dir + archive_name + ".pkl";
+  writer.writeRecord(fname, data.data(), data.size());
+
+  // serialize all the captured run-time class types
+  //  for (const c10::ClassTypePtr& wroteType : memoizedClassTypes) {
+  //    convertNamedType(wroteType);
+  //  }
```
Remove

raziel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -629,12 +627,12 @@ void ScriptModuleSerializer::writeByteCode(
   BackendDebugInfoRecorder debug_info_recorder;
-  elements.emplace_back(
-      static_cast<int64_t>(caffe2::serialize::kProducedBytecodeVersion));
+  int64_t version_to_write = caffe2::serialize::kProducedBytecodeVersion;
```
```suggestion
  cosnt int64_t version_to_write = caffe2::serialize::kProducedBytecodeVersion;
```

or just use kProducedBytecodeVersion directly

iseeyuan replied:
Yeah I should use kProducedBytecodeVersion directly. It's remainders from last commit. Thanks!

raziel commented on torch/csrc/jit/serialization/export_module.cpp:
```diff
@@ -849,5 +847,17 @@ std::vector<std::string> export_opnames(const script::Module& m) {
+// to control if instructions for bytecode default inputs are emitted
+// or not. It's the major difference between bytecode v5 and v6.
+thread_local bool emitBytecodeDefaultInputs =
+    caffe2::serialize::kProducedBytecodeVersion <= 5 ? true : false;
```
Why not just set this to false?
If the only intention for the condition is to signal this is new logic for v6 then the comment should suffice.

Otherwise I think it looks odd to have a condition based on a constant.

iseeyuan replied:
Makes sense. I'll set it false.

raziel reviewed (approved):
Thanks!

There's just a few comments to address.

gmagogsfm commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -211,6 +212,152 @@ std::stringstream backport_v5_to_v4(std::stringstream& input_model_stream) {
+  //  }
+}
+
+std::stringstream backport_v6_to_v5(std::stringstream& input_model_stream) {
```
Implementation looks good, however I have a high level question: why do we backport byte code and constant table when we can re-generate an entirely new archive that is v5 compatible?

iseeyuan replied:
Sorry I missed this question. The purpose for backport is for forward compatibility: to deploy a new model (generated by new code) to old runtime. The reason we use backport function, is to keep one unified serialization version for current code to avoid confusion. If we re-generate an entirely new archive for v5, that means that we maintain serialization flows to both v6 and v5 in the main code. Another reason is to export to even older versions, that the backport functions can be called in a while loop. In general the backport function should be just format change. 

It's special case for this v6 update, where the bytecode must be remitted to backport, that changing format of bytecode is not possible.

gmagogsfm replied:
>  If we re-generate an entirely new archive for v5, that means that we maintain serialization flows to both v6 and v5 in the main code. 

By maintaining backport code, it is pretty much the same as maintaining both v5 and v6 export in main code.

> Another reason is to export to even older versions, that the backport functions can be called in a while loop. In general the backport function should be just format change.

Do we need to backport to older than v5 (like v4)?

gmagogsfm replied:
To be clear, I think current approach works too, just curious why it is done this way.

iseeyuan committed:
Update on "Enable implicit operator versioning via number of arguments"


Enable implicit operator versioning via number of arguments from Mobile.
1. By default, TS doesn't emit instructions for tailing default args and the provided number of specified args is serialized to bytecode. From interpreter the default values are fetched from operator schema. The implementation has been landed in #56845. Please refer to #56845 for details.
2. Since there is bytecode schema change, the bytecode version is bumped from 5 to 6.
3. The corresponding backport function is provided, for forward compatibility use. Note that because there is instruction change, a global flag is used as the switch to control the two versions.

Differential Revision: [D28789746](https://our.internmc.facebook.com/intern/diff/D28789746)

[ghstack-poisoned]

iseeyuan commented:
@iseeyuan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28789746).

iseeyuan committed:
Update on "Enable implicit operator versioning via number of arguments"


Enable implicit operator versioning via number of arguments from Mobile.
1. By default, TS doesn't emit instructions for tailing default args and the provided number of specified args is serialized to bytecode. From interpreter the default values are fetched from operator schema. The implementation has been landed in #56845. Please refer to #56845 for details.
2. Since there is bytecode schema change, the bytecode version is bumped from 5 to 6.
3. The corresponding backport function is provided, for forward compatibility use. Note that because there is instruction change, a global flag is used as the switch to control the two versions.

Differential Revision: [D28789746](https://our.internmc.facebook.com/intern/diff/D28789746)

[ghstack-poisoned]

iseeyuan commented:
@iseeyuan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28789746).
